### ISSUE TITLE
Simplify rule for attribute values on Extension Instructions used to invoke named templates

### ISSUE THREAD

**Issue by pgfearo** [qt4cg/qtspecs#92]
Regarding the rule in the current proposal for [Invoking Named Templates with Extension Instructions](https://qt4cg.org/branch/master/xslt-40/Overview.html#invoking-named-templates):

> The way in which attribute values are handled depends on the type declaration of the template parameter...

I have some problems with this dependency on parameter type (to control whether value is an AVT or XPath expression):

1. In many cases, a `xs:string` or `xs:boolean` type passed as a param will be a variable reference so a coder needs to enter `name="{$myName}"` instead of `name="$myName"` in their XSLT editor.
2. If passing a literal `xs:string` type, the syntax: `name="first"` would be easy for a human reader to misinterpret as a `NameTest` instead of a `StringLiteral`.
3. The dependency on param type means more effort (and thus poorer performance) for a tokenizer or syntax-highlighter as it may need to get type information from included/imported XSLT stylesheet modules or from extension elements declared later in the same XSLT module.

The third point above is most important from my viewpoint as maintainer of an XSLT editor, but I believe the first two points are also valid.

For these reasons, I propose that: all attribute-values on extension instructions used to invoke named templates are treated as XPath expressions.

**Comment by pgfearo** [comment]
@yamahito I agree that in the proposed specification there is indeed contradiction in the documentation and in the examples. 

However, in the [Saxon documentation](https://www.saxonica.com/documentation11/index.html#!v4extensions/xslt-syntax-extensions/ext-ins-call-template) though there is no such contradiction. The Saxon 11.2 behaviour is consistent with the examples.

As an XSLT developer I pass variable/parameter values much more frequently than string-literals. Hence my preference to use:

`<log:message message="$greeting"/`

Instead of:

`<log:message message="{$greeting}"/`

For the rare occasions where I can hard-code a string-literal the extra nested quotes would (in my view) not be too burdensome:

`<log:message message="'good morning'"/`

As a hack I guess I could always declare the parameter as `xs:string+` to get the desired behaviour:
```xml 
<xsl:template name="log:message">
    <xsl:param name="message" as="xs:string+"/>
    ...
</xsl:template>
```
I will delay making a decision on how I support the type-dependant aspect of this XSLT 4.0 feature in VS Code until there's been more discussion on this.  So for now, extension instructions will have their attributes recognised as AVT's regardless of parameter type.

The following screenshot of XSLT than runs with Saxon11 shows the problem this causes in the editor:

![Screenshot 2022-03-13 at 09 29 59](https://user-images.githubusercontent.com/2734234/158053387-41db423d-3bb3-403d-857a-a5772db1a148.png)

1. `$greetingList` on line 12 is marked as unused (greyed out)
2. `$greetingList` on line 16 is wrongly highlighted as a string-literal instead of a variable (orange instead of white)

**Comment by michaelhkay** [comment]
The purpose of the rules as written is to make these extension functions behave as far as possible in the same way as typical built-in instructions.

Most boolean parameters to existing instructions expect literal yes/no values, for example `streamable=yes|no`. In at least one case (`xsl:message terminate="yes|no"`, the value can be supplied as an AVT (and we changed these parameters to accept true|false|0|1 to make this convenient). Users would not expect to pass such parameters as XPath expressions.

String-valued parameters to existing instructions generally expect either literal values (for example `validation="strict"`, or AVTs (`regex="{$regex"`). Users would not expect to double the quotation marks on such parameters (`validation="'strict'") which would be needed if the value is an XPath expression.

I think your arguments (1) and (2) apply equally well to the conventions followed for existing XSLT instructions, and I think it's more important to be consistent with these conventions than to improve on them.

As for argument (3), I have learned over the years that designing a specification for the benefit of implementors rather than users is rarely a good idea. Syntax-directed editors do have a problem with XSLT modules containing references to other modules that might not be directly imported or included; I don't feel we are making that problem any worse.
